File: python/src/agent_squad/retrievers/dakera_retriever.py
import os
from dataclasses import dataclass
from typing import Any, Dict, List, Optional

from agent_squad.retrievers import Retriever


@dataclass
class DakeraRetrieverOptions:
    """Options for the Dakera retriever.

    Attributes:
        namespace: The Dakera namespace to query.
        api_key: Dakera API key (a ``dk-...`` token). Falls back to the
            ``DAKERA_API_KEY`` environment variable when not set.
        url: Base URL of the Dakera server. Falls back to the ``DAKERA_URL``
            environment variable and finally to ``http://localhost:3000``.
        top_k: Maximum number of results to return.
        filter: Optional Dakera metadata filter applied to the query.
    """

    namespace: str
    api_key: Optional[str] = None
    url: str = ""
    top_k: int = 10
    filter: Optional[Dict] = None


class DakeraRetriever(Retriever):
    """Retriever backed by a self-hosted `Dakera <https://dakera.ai>`_ memory server.

    Uses Dakera's text-query API (server-side embedding) to fetch the most
    relevant documents for a query, which agents can use as retrieval-augmented
    context.
    """

    def __init__(self, options: DakeraRetrieverOptions):
        super().__init__(options)
        self.options = options

        if not self.options.namespace:
            raise ValueError("namespace is required in options")

        api_key = self.options.api_key or os.getenv("DAKERA_API_KEY")
        if not api_key:
            raise ValueError("api_key is required (set it in options or the DAKERA_API_KEY env var)")
        url = self.options.url or os.getenv("DAKERA_URL") or "http://localhost:3000"

        try:
            from dakera import AsyncDakeraClient
        except ImportError as exc:
            raise ImportError(
                "The 'dakera' package is required to use DakeraRetriever. "
                "Install it with: pip install 'agent-squad[dakera]'"
            ) from exc

        self.client = AsyncDakeraClient(base_url=url, api_key=api_key)

    async def retrieve(
        self,
        text: str,
        top_k: Optional[int] = None,
        metadata_filter: Optional[Dict] = None,
    ) -> List[Any]:
        """Retrieve the documents most relevant to ``text`` from Dakera.

        Returns the list of ``TextSearchResult`` objects (``.id``, ``.score``,
        ``.text``, ``.metadata``) from the Dakera text query.
        """
        if not text:
            raise ValueError("Input text is required for retrieve")

        response = await self.client.query_text(
            self.options.namespace,
            text=text,
            top_k=top_k or self.options.top_k,
            filter=metadata_filter if metadata_filter is not None else self.options.filter,
        )
        return response.results

    async def retrieve_and_combine_results(
        self,
        text: str,
        top_k: Optional[int] = None,
        metadata_filter: Optional[Dict] = None,
    ) -> str:
        """Retrieve results for ``text`` and combine their text into one string."""
        results = await self.retrieve(text, top_k, metadata_filter)
        return self.combine_retrieval_results(results)

    async def retrieve_and_generate(self, text: str) -> Any:
        """Not supported: Dakera is a retrieval-only backend (no generation)."""
        raise NotImplementedError(
            "DakeraRetriever does not support retrieve_and_generate; use retrieve or retrieve_and_combine_results."
        )

    @staticmethod
    def combine_retrieval_results(results: List[Any]) -> str:
        """Join the ``text`` field of each result into a single newline-separated string."""
        return "\n".join(result.text for result in results if result is not None and getattr(result, "text", None))

File: typescript/src/retrievers/DakeraRetriever.ts
import { Retriever } from "./retriever";
// Type-only import: erased at compile time, so it adds no runtime dependency.
// `@dakera-ai/dakera` is an optional peer dependency, loaded lazily in the
// constructor below so `agent-squad` installs and imports without it.
import type { DakeraClient, FilterExpression } from "@dakera-ai/dakera";

/**
 * Interface defining the options for DakeraRetriever.
 */
export interface DakeraRetrieverOptions {
  /** The Dakera namespace to query. */
  namespace: string;
  /** Dakera API key (a `dk-...` token). Falls back to the `DAKERA_API_KEY` env var. */
  apiKey?: string;
  /** Base URL of the Dakera server. Falls back to `DAKERA_URL`, then `http://localhost:3000`. */
  url?: string;
  /** Maximum number of results to return. Defaults to 10. */
  topK?: number;
  /** Optional Dakera metadata filter applied to the query. */
  filter?: FilterExpression;
}

/**
 * Retriever backed by a self-hosted [Dakera](https://dakera.ai) memory server.
 *
 * Uses Dakera's text-query API (server-side embedding) to fetch the most relevant
 * documents for a query, which agents can use as retrieval-augmented context.
 * Extends the base {@link Retriever} class.
 *
 * `@dakera-ai/dakera` is an optional peer dependency: it is `require`d lazily in
 * the constructor, so installing `agent-squad` never pulls it in unless this
 * retriever is actually used. A clear error is thrown if it is missing.
 */
export class DakeraRetriever extends Retriever {
  private client: DakeraClient;
  protected options: DakeraRetrieverOptions;

  /**
   * Constructor for DakeraRetriever.
   * @param options - Configuration options for the retriever.
   */
  constructor(options: DakeraRetrieverOptions) {
    super(options);
    this.options = options;

    if (!this.options.namespace) {
      throw new Error("namespace is required in options");
    }

    const apiKey = this.options.apiKey ?? process.env.DAKERA_API_KEY;
    if (!apiKey) {
      throw new Error(
        "apiKey is required (set it in options or the DAKERA_API_KEY env var)",
      );
    }
    const baseUrl = this.options.url || process.env.DAKERA_URL || "http://localhost:3000";

    // Lazily load the optional peer dependency. Keeping the require here (rather
    // than a top-level import) means the SDK is only needed when this retriever
    // is instantiated, not by everyone who installs agent-squad.
    let DakeraClientCtor: typeof import("@dakera-ai/dakera").DakeraClient;
    try {
      // eslint-disable-next-line @typescript-eslint/no-require-imports
      ({ DakeraClient: DakeraClientCtor } = require("@dakera-ai/dakera"));
    } catch {
      throw new Error(
        "DakeraRetriever requires the optional peer dependency '@dakera-ai/dakera'. " +
          "Install it with: npm install @dakera-ai/dakera",
      );
    }

    this.client = new DakeraClientCtor({ baseUrl, apiKey });
  }

  /**
   * Retrieve the documents most relevant to `text` from Dakera.
   * @param text - The query text.
   * @returns The text query results (each with `id`, `score` and `text`).
   */
  public async retrieve(text: string): Promise<any[]> {
    if (!text) {
      throw new Error("Input text is required for retrieve");
    }

    const response = await this.client.queryText(this.options.namespace, text, {
      topK: this.options.topK ?? 10,
      filter: this.options.filter,
    });
    return response.results;
  }

  /**
   * Retrieve results for `text` and combine their text into a single string.
   * @param text - The query text.
   * @returns The combined result text, newline-separated.
   */
  public async retrieveAndCombineResults(text: string): Promise<string> {
    const results = await this.retrieve(text);
    return this.combineRetrievalResults(results);
  }

  /**
   * Not supported: Dakera is a retrieval-only backend (no generation).
   * @throws Always, directing callers to `retrieve` / `retrieveAndCombineResults`.
   */
  public async retrieveAndGenerate(_text: string): Promise<any> {
    throw new Error(
      "DakeraRetriever does not support retrieveAndGenerate; " +
        "use retrieve or retrieveAndCombineResults.",
    );
  }

  /**
   * Combine the `text` field of each result into a single newline-separated string.
   * @param results - The Dakera text query results.
   * @returns The combined text.
   */
  private combineRetrievalResults(results: Array<{ text?: string }>): string {
    return results
      .filter((result) => result && typeof result.text === "string")
      .map((result) => result.text as string)
      .join("\n");
  }
}

